File: 2019/intmachine.go
package main

import "fmt"

// IntMachine is a virtual machine that runs an intcode
type IntMachine struct {
	programCounter int
	memory         []int
	inputChannel   chan int
	outputChannel  chan int
}

func (machine *IntMachine) getInstruction() int {
	return machine.memory[machine.programCounter]
}

func (machine *IntMachine) getPointer(mode, pointer int) int {
	switch mode {
	case 0:
		return machine.memory[pointer] // parameter mode
	case 1:
		return pointer // immediate mode
	default:
		return -1
	}
}

func parseInstruction(opcode int) (int, int, int, int) {
	op := opcode % 100
	mode1 := opcode / 100 % 10
	mode2 := opcode / 1000 % 10
	mode3 := opcode / 10000 % 10
	return op, mode1, mode2, mode3
}

func (machine *IntMachine) addition(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	aPtr := machine.getPointer(mode1, pc+1)
	bPtr := machine.getPointer(mode2, pc+2)
	writePtr := machine.getPointer(mode3, pc+3)
	machine.memory[writePtr] = machine.memory[aPtr] + machine.memory[bPtr]
	machine.programCounter += 4
}

func (machine *IntMachine) multiplication(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	aPtr := machine.getPointer(mode1, pc+1)
	bPtr := machine.getPointer(mode2, pc+2)
	writePtr := machine.getPointer(mode3, pc+3)
	machine.memory[writePtr] = machine.memory[aPtr] * machine.memory[bPtr]
	machine.programCounter += 4
}

func (machine *IntMachine) input(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	writePtr := machine.getPointer(mode1, pc+1)
	machine.memory[writePtr] = <-machine.inputChannel
	machine.programCounter += 2
}

func (machine *IntMachine) output(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	readPtr := machine.getPointer(mode1, pc+1)
	machine.outputChannel <- machine.memory[readPtr]
	machine.programCounter += 2
}

func (machine *IntMachine) jumpIfTrue(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	boolPtr := machine.getPointer(mode1, pc+1)
	jumpPtr := machine.getPointer(mode2, pc+2)
	if machine.memory[boolPtr] == 0 {
		machine.programCounter += 3
	} else {
		machine.programCounter = machine.memory[jumpPtr]
	}
}

func (machine *IntMachine) jumpIfFalse(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	boolPtr := machine.getPointer(mode1, pc+1)
	jumpPtr := machine.getPointer(mode2, pc+2)
	if machine.memory[boolPtr] == 0 {
		machine.programCounter = machine.memory[jumpPtr]
	} else {
		machine.programCounter += 3
	}
}

func (machine *IntMachine) lessThan(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	lowPtr := machine.getPointer(mode1, pc+1)
	highPtr := machine.getPointer(mode2, pc+2)
	writePtr := machine.getPointer(mode3, pc+3)
	if machine.memory[lowPtr] < machine.memory[highPtr] {
		machine.memory[writePtr] = 1
	} else {
		machine.memory[writePtr] = 0
	}
	machine.programCounter += 4
}

func (machine *IntMachine) equals(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	aPtr := machine.getPointer(mode1, pc+1)
	bPtr := machine.getPointer(mode2, pc+2)
	writePtr := machine.getPointer(mode3, pc+3)
	if machine.memory[aPtr] == machine.memory[bPtr] {
		machine.memory[writePtr] = 1
	} else {
		machine.memory[writePtr] = 0
	}
	machine.programCounter += 4
}

func (machine *IntMachine) halt(mode1, mode2, mode3 int) {
	pc := machine.programCounter
	fmt.Printf("Program halted at: %d with exit code: %d\n", pc, 1)
}

var opcodes = map[int]func(*IntMachine, int, int, int){
	1:  (*IntMachine).addition,
	2:  (*IntMachine).multiplication,
	3:  (*IntMachine).input,
	4:  (*IntMachine).output,
	5:  (*IntMachine).jumpIfTrue,
	6:  (*IntMachine).jumpIfFalse,
	7:  (*IntMachine).lessThan,
	8:  (*IntMachine).equals,
	99: (*IntMachine).halt}

func (machine *IntMachine) run() {
	for {
		op, mode1, mode2, mode3 := parseInstruction(machine.getInstruction())
		opcodes[op](machine, mode1, mode2, mode3)
		if op == 99 {
			break // halt program
		}
	}
}

func memClone(a []int) []int {
	b := make([]int, len(a))
	copy(b, a)
	return b
}

func machineLoop(code []int, number int) []IntMachine {
	machines := make([]IntMachine, number)

	// create machines
	for i := 0; i < number; i++ {
		machine := IntMachine{0, memClone(code), nil, nil}
		machines = append(machines, machine)
	}

	// link machines
	for i := 0; i < number; i++ {
		ch := make(chan int)
		machines[i].outputChannel = ch
		machines[i+1%number].inputChannel = ch
	}
	return machines
}

func permutations(arr []int) [][]int {
	res := [][]int{}
	permute(arr, len(arr))
	return res
}

func permute(arr []int, n int) {
	if n == 1 {
		tmp := make([]int, len(arr))
		copy(tmp, arr)
		res := append(res, tmp)
	} else {
		for i := 0; i < n; i++ {
			permute(arr, n-1)
			if n%2 == 1 {
				tmp := arr[i]
				arr[i] = arr[n-1]
				arr[n-1] = tmp
			} else {
				tmp := arr[0]
				arr[0] = arr[n-1]
				arr[n-1] = tmp
			}
		}
	}
}

func main() {
	data := []int{3, 8, 1001, 8, 10, 8, 105, 1, 0, 0, 21, 42, 55, 76, 89, 114, 195, 276, 357, 438, 99999, 3, 9, 1001, 9, 3, 9, 1002, 9, 3, 9, 1001, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 99, 3, 9, 102, 2, 9, 9, 101, 5, 9, 9, 4, 9, 99, 3, 9, 102, 3, 9, 9, 101, 5, 9, 9, 1002, 9, 2, 9, 101, 4, 9, 9, 4, 9, 99, 3, 9, 102, 5, 9, 9, 1001, 9, 3, 9, 4, 9, 99, 3, 9, 1001, 9, 4, 9, 102, 5, 9, 9, 1001, 9, 5, 9, 1002, 9, 2, 9, 101, 2, 9, 9, 4, 9, 99, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 1001, 9, 1, 9, 4, 9, 3, 9, 1001, 9, 2, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 1001, 9, 2, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 1001, 9, 2, 9, 4, 9, 99, 3, 9, 1001, 9, 2, 9, 4, 9, 3, 9, 101, 2, 9, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 99, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 101, 2, 9, 9, 4, 9, 99, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 1001, 9, 2, 9, 4, 9, 3, 9, 101, 2, 9, 9, 4, 9, 3, 9, 1001, 9, 1, 9, 4, 9, 3, 9, 101, 2, 9, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 1001, 9, 1, 9, 4, 9, 3, 9, 1001, 9, 2, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 1001, 9, 1, 9, 4, 9, 99, 3, 9, 1001, 9, 1, 9, 4, 9, 3, 9, 101, 1, 9, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 101, 2, 9, 9, 4, 9, 3, 9, 1001, 9, 1, 9, 4, 9, 3, 9, 1002, 9, 2, 9, 4, 9, 3, 9, 102, 2, 9, 9, 4, 9, 3, 9, 101, 2, 9, 9, 4, 9, 99}
	machines := machineLoop(data, 5)

}

File: 2019/intmachine.py
class IntMachine():
    def __init__(self, code, cin, cout):
        self.pointer = 0
        self.read_head = 0
        self.write_head = 0
        self.code = code
        self.cin = cin
        self.cout = cout
        self.opcodes = {
            1: self.addition,
            2: self.multiplication,
            3: self.take,
            4: self.give,
            5: self.jump_if_true,
            6: self.jump_if_false,
            7: self.less_than,
            8: self.equals,
            99: self.end
        }

    def get(self, mode, param):
        if mode == 0: # position mode
            return self.code[param]
        elif mode == 1: # immediate mode
            return param

    def parse(self, opcode):
        op = opcode % 100
        mode1 = opcode // 100 % 10
        mode2 = opcode // 1000 % 10
        mode3 = opcode // 10000 % 10
        return (op, (mode1, mode2, mode3))

    def run(self):
        while True:
            (op, modes) = self.parse(self.code[self.pointer])
            self.opcodes[op](modes, self.pointer)
            if op == 99: break # halt program, but end func runs first

    def addition(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        b = self.get(m2, self.code[ptr+2])
        c = self.code[ptr+3]
        self.code[c] = a + b
        self.pointer += 4

    def multiplication(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        b = self.get(m2, self.code[ptr+2])
        c = self.code[ptr+3]
        self.code[c] = a * b
        self.pointer += 4

    def take(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.code[ptr+1]
        self.code[a] = self.cin[self.read_head]
        self.read_head += 1
        self.pointer += 2

    def give(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        self.cout.append(a)
        self.write_head += 1
        self.pointer += 2

    def jump_if_true(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        b = self.get(m2, self.code[ptr+2])
        if a == 0:
            self.pointer += 3
        else:
            self.pointer = b

    def jump_if_false(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        b = self.get(m2, self.code[ptr+2])
        if a == 0:
            self.pointer = b
        else:
            self.pointer += 3

    def less_than(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        b = self.get(m2, self.code[ptr+2])
        c = self.code[ptr+3]
        if a < b:
            self.code[c] = 1
        else:
            self.code[c] = 0
        self.pointer += 4

    def equals(self, modes, ptr):
        (m1, m2, m3) = modes
        a = self.get(m1, self.code[ptr+1])
        b = self.get(m2, self.code[ptr+2])
        c = self.code[ptr+3]
        if a == b:
            self.code[c] = 1
        else:
            self.code[c] = 0
        self.pointer += 4

    def end(self, modes, ptr):
        print(f"Program halted at: {ptr} with diagnostic code: {self.cout[-1]}")


if __name__ == "__main__":
    data = "3,225,1,225,6,6,1100,1,238,225,104,0,1102,88,66,225,101,8,125,224,101,-88,224,224,4,224,1002,223,8,223,101,2,224,224,1,224,223,223,1101,87,23,225,1102,17,10,224,101,-170,224,224,4,224,102,8,223,223,101,3,224,224,1,223,224,223,1101,9,65,225,1101,57,74,225,1101,66,73,225,1101,22,37,224,101,-59,224,224,4,224,102,8,223,223,1001,224,1,224,1,223,224,223,1102,79,64,225,1001,130,82,224,101,-113,224,224,4,224,102,8,223,223,1001,224,7,224,1,223,224,223,1102,80,17,225,1101,32,31,225,1,65,40,224,1001,224,-32,224,4,224,102,8,223,223,1001,224,4,224,1,224,223,223,2,99,69,224,1001,224,-4503,224,4,224,102,8,223,223,101,6,224,224,1,223,224,223,1002,14,92,224,1001,224,-6072,224,4,224,102,8,223,223,101,5,224,224,1,223,224,223,102,33,74,224,1001,224,-2409,224,4,224,1002,223,8,223,101,7,224,224,1,223,224,223,4,223,99,0,0,0,677,0,0,0,0,0,0,0,0,0,0,0,1105,0,99999,1105,227,247,1105,1,99999,1005,227,99999,1005,0,256,1105,1,99999,1106,227,99999,1106,0,265,1105,1,99999,1006,0,99999,1006,227,274,1105,1,99999,1105,1,280,1105,1,99999,1,225,225,225,1101,294,0,0,105,1,0,1105,1,99999,1106,0,300,1105,1,99999,1,225,225,225,1101,314,0,0,106,0,0,1105,1,99999,107,677,677,224,1002,223,2,223,1006,224,329,101,1,223,223,108,677,677,224,1002,223,2,223,1005,224,344,101,1,223,223,1007,677,677,224,1002,223,2,223,1006,224,359,101,1,223,223,1107,226,677,224,1002,223,2,223,1006,224,374,1001,223,1,223,8,677,226,224,1002,223,2,223,1006,224,389,101,1,223,223,1108,677,677,224,1002,223,2,223,1005,224,404,1001,223,1,223,7,226,226,224,1002,223,2,223,1006,224,419,101,1,223,223,1107,677,677,224,1002,223,2,223,1005,224,434,101,1,223,223,107,226,226,224,102,2,223,223,1005,224,449,101,1,223,223,107,677,226,224,1002,223,2,223,1006,224,464,1001,223,1,223,8,226,677,224,102,2,223,223,1006,224,479,1001,223,1,223,108,677,226,224,102,2,223,223,1005,224,494,1001,223,1,223,1108,677,226,224,1002,223,2,223,1005,224,509,1001,223,1,223,1107,677,226,224,1002,223,2,223,1005,224,524,101,1,223,223,1008,226,226,224,1002,223,2,223,1006,224,539,101,1,223,223,1008,226,677,224,1002,223,2,223,1005,224,554,1001,223,1,223,7,226,677,224,1002,223,2,223,1005,224,569,101,1,223,223,1007,677,226,224,1002,223,2,223,1006,224,584,1001,223,1,223,7,677,226,224,102,2,223,223,1006,224,599,101,1,223,223,1007,226,226,224,102,2,223,223,1006,224,614,101,1,223,223,1008,677,677,224,1002,223,2,223,1006,224,629,101,1,223,223,108,226,226,224,102,2,223,223,1006,224,644,101,1,223,223,1108,226,677,224,1002,223,2,223,1005,224,659,101,1,223,223,8,226,226,224,1002,223,2,223,1005,224,674,101,1,223,223,4,223,99,226"
    code = [int(x) for x in data.split(",")]
    cin = [5]
    cout = []
    machine = IntMachine(code, cin, cout)
    machine.run()
    print(cout)